test/jdk/java/nio/channels/DatagramChannel/SRTest.java

```diff
@@ -24,184 +24,220 @@
+        CompletableFuture<Void> f2 = CompletableFuture.runAsync(reader, e);
+        wait(f1, f2);
     }
 
-    public interface Sprintable extends Runnable {
-        public void throwException() throws Exception;
+    // Exit with CompletionException if any passed futures complete exceptionally
+    private static void wait(CompletableFuture<?>... futures) throws CompletionException {
+        CompletableFuture<?> future = CompletableFuture.allOf(futures);
+        Stream.of(futures)
+                .forEach(f -> f.exceptionally(ex -> {
+                    future.completeExceptionally(ex);
+                    return null;
+                }));
+        future.join();
     }
 
-    public static class ClassicWriter implements Sprintable {
+    public static class ClassicWriter implements Runnable, AutoCloseable {
+        final DatagramSocket ds;
         final int port;
-        Exception e = null;
 
-        ClassicWriter(int port) {
+        ClassicWriter(int port) throws SocketException {
             this.port = port;
-        }
-
-        public void throwException() throws Exception {
-            if (e != null)
-                throw e;
+            this.ds = new DatagramSocket();
         }
 
         public void run() {
             try {
-                DatagramSocket ds = new DatagramSocket();
                 String dataString = "hello";
                 byte[] data = dataString.getBytes();
-                InetAddress address = InetAddress.getLocalHost();
+                InetAddress address = InetAddress.getLoopbackAddress();
                 DatagramPacket dp = new DatagramPacket(data, data.length,
                                                        address, port);
                 ds.send(dp);
                 Thread.sleep(50);
                 ds.send(dp);
-            } catch (Exception ex) {
-                e = ex;
+            } catch (Exception e) {
+                log.println("ClassicWriter [" + ds.getLocalAddress() + "]");
+                throw new RuntimeException("ClassicWriter threw exception: " + e);
+            } finally {
+                log.println("ClassicWriter finished");
             }
         }
+
+        @Override
+        public void close() throws IOException {
+            ds.close();
+        }
     }
 
-    public static class NioWriter implements Sprintable {
+    public static class NioWriter implements Runnable, AutoCloseable {
+        final DatagramChannel dc;
         final int port;
-        Exception e = null;
 
-        NioWriter(int port) {
+        NioWriter(int port) throws IOException {
+            this.dc = DatagramChannel.open();
             this.port = port;
         }
 
-        public void throwException() throws Exception {
-            if (e != null)
-                throw e;
-        }
-
         public void run() {
             try {
                 DatagramChannel dc = DatagramChannel.open();
```

Comment:
This line needs to be removed since `dc` is now created in the constructor.

Reply:
Dang, that a bad one. Thank you for spotting! Didn't cause any test failures but of course very wrong.